File: julia/roman-numerals/roman-numerals.jl
function to_roman(number)
    (number > 0 && number < 4000) || error("$number cannot be converted to roman")

    parts = (roman_digit(d, roman_infos[i]) for (i, d) in enumerate(digits(number)))
    parts |> Iterators.reverse |> join
end

struct RomanInfo
    one::Char
    five::Char
    ten::Char
end

const roman_infos = [
    RomanInfo('I', 'V', 'X'),
    RomanInfo('X', 'L', 'C'),
    RomanInfo('C', 'D', 'M'),
    RomanInfo('M', '!', '!')
]

function roman_digit(digit, infos::RomanInfo)
    if digit < 4
        infos.one ^ digit
    elseif digit == 4
        string(infos.one, infos.five)
    elseif digit < 9
        string(infos.five, infos.one ^ (digit - 5))
    else
        string(infos.one, infos.ten)
    end
end

File: ruby/roman-numerals/roman_numerals.rb
class Integer
  def to_roman
    raise RuntimeError, "Roman numerals can only express 1..3999" unless (1..3999) === self
    digits.each.with_index.map { |d, i| digit_to_roman(d, NUMERAL_INFOS[i]) }.reverse.join
  end

  private

  NumeralInfo = Struct.new(:one, :five, :ten)

  NUMERAL_INFOS = [
    NumeralInfo.new("I", "V", "X"),
    NumeralInfo.new("X", "L", "C"),
    NumeralInfo.new("C", "D", "M"),
    NumeralInfo.new("M")
  ]

  def digit_to_roman(digit, numeral_info)
    case digit
    when 0...4
      numeral_info.one * digit
    when 4
      numeral_info.one + numeral_info.five
    when 5...9
      numeral_info.five + (numeral_info.one * (digit - 5))
    when 9
      numeral_info.one + numeral_info.ten
    else
      ""
    end
  end
end
